daemon/build.go

```diff
@@ -104,13 +105,19 @@ func (daemon *Daemon) pullForBuilder(ctx context.Context, name string, authConfi
 // Every call to GetImageAndReleasableLayer MUST call releasableLayer.Release() to prevent
 // leaking of layers.
 func (daemon *Daemon) GetImageAndReleasableLayer(ctx context.Context, refOrID string, opts backend.GetImageAndLayerOptions) (builder.Image, builder.ReleaseableLayer, error) {
-	if !opts.ForcePull {
-		image, _ := daemon.GetImage(refOrID)
+	id, _ := daemon.GetImageID(refOrID)
+	refIsID := id.String() == refOrID // detect if ref is an ID to skip pulling
```

Comment:
I'm not sure how safe that would be. Prefix ID and references are ambiguous. If the prefix matches local image it doesn't really mean that pulling would never be needed. Probably ok to assume that `--pull` and ambiguous explicit local IDs don't work together, if not then security suffers a bit. In this case full ID is automatically taken from build stage, so the user never sets it.